helita/sim/bifrost.py

```diff
@@ -329,44 +360,196 @@ def _get_simple_var_xy(self, var, order='F', mode='r'):
-            p = self.get_var('p' + var[1])
-            if getattr(self, 'n' + var[1]) < 5:
-                return p / self.r   # do not recentre for 2D cases (or close)
+            if not hasattr(self, 'p'+ var[1]): setattr( self, 'p' + var[1],self.get_var('p' + var[1],self.snap))
+            if not hasattr(self, 'r'): self.r = self.variables['r']= self.get_var('r',self.snap)
+            self.boundcut('r',self.r)
+            self.boundcut('p'+ var[1],getattr(self,'p'+ var[1]))
+            if (getattr(self, 'n' + var[1]) < 5):
+                return getattr(self,'p'+ var[1]) / self.r   # do not recentre for 2D cases (or close)
             else:  # will call xdn, ydn, or zdn to get r at cell faces
-                return p / getattr(cstagger, var[1] + 'dn')(self.r)
+                rdt = self.r.dtype
+                cs.init_stagger(self.nz, self.dx, self.dy, self.z.astype(rdt), self.zdn.astype(rdt), self.dzidzup.astype(rdt), self.dzidzdn.astype(rdt))
+                return getattr(self,'p'+ var[1]) / getattr(cs, var[1] + 'dn')(self.r)
         elif var == 'ee':   # internal energy
             if not hasattr(self, 'e'):
-                self.e = self.variables['e'] = self.get_var('e')
+                self.e = self.variables['e'] = self.get_var('e',self.snap)
+            if not hasattr(self, 'r'): self.r = self.variables['r'] = self.get_var('r',self.snap)
             return self.e / self.r
         elif var == 's':   # entropy?
             if not hasattr(self, 'p'):
-                self.p = self.variables['p'] = self.get_var('p')
-            return np.log(self.p) - 1.667 * np.log(self.r)
-        elif var in ['modb', 'modp']:   # total magnetic field
-            v = var[3]
+                self.p = self.variables['p'] = self.get_var('p',self.snap)
+            return np.log(self.p) - self.params['gamma'] * np.log(self.r)
+        elif var[1] in ['2']:   # total magnetic field
+            v = var[0]
             if v == 'b':
                 if not self.do_mhd:
                     raise ValueError("No magnetic field available.")
             rdt = self.r.dtype
-            cstagger.init_stagger(self.nzb, self.z.astype(rdt),
-                                  self.zdn.astype(rdt))
-            result = cstagger.xup(getattr(self, v+'x')) ** 2
-            result += cstagger.yup(getattr(self, v+'y')) ** 2
-            result += cstagger.zup(getattr(self, v+'z')) ** 2
-            return np.sqrt(result)
+            cs.init_stagger(self.nz, self.dx, self.dy, self.z.astype(rdt), self.zdn.astype(rdt), self.dzidzup.astype(rdt), self.dzidzdn.astype(rdt))
+            result = cs.xup(getattr(self, v+'x')) ** 2
+            result += cs.yup(getattr(self, v+'y')) ** 2
+            result += cs.zup(getattr(self, v+'z')) ** 2
+            return result
+        elif (len(var) > 2):
+            if var in ['ixc', 'iyc', 'izc'] or var in ['exc', 'eyc', 'ezc']:
+                p = self.variables[var[0:2] +'c'] = self.get_var(var[0:2],self.snap)
```

Comment:
There must be another way for this. Calling `self.get_var` from inside `self._get_composite_var` could end up in an infinite loop. Please don't do it.